lib/ddtrace/contrib/rack/rum_injection.rb

```diff
@@ -0,0 +1,273 @@
+          # TODO: glob performance may be worse than regex
+          return false unless configuration[:rum_cached_pages].none? do |page_glob|
+            File.fnmatch(page_glob, env['PATH_INFO']) if env['PATH_INFO']
+          end
+
+          return false unless headers
+
+          # first check Surrogate-Controle, which fastly uses
+          # indicates a cdn cache if Surrogate-Control max-age>0,
+          # otherwise check cache-control, then expiry to determine if there is browser cache
+          return false if surrogate_cache?(headers)
+
+          # then check Cache-Control
+          if headers.key?(CACHE_CONTROL_HEADER) && !headers[CACHE_CONTROL_HEADER].nil?
+            # s-maxage gets priority over max-age since s-maxage sits at cdn level
+            # cdn cache if s-maxage >0,
+            # otherwise check max-age, (no-store|no-cache|private) or expires to determine if therre is browser cache
+            return false if server_cache?(headers)
+
+            # then check max-age
+            if headers[CACHE_CONTROL_HEADER].include?(MAX_AGE)
+              # only not cached if max-age is 0
+              return headers[CACHE_CONTROL_HEADER].include?(MAX_AGE_ZERO)
+            end
+
+            # not cached if marked no-store, no-cache, or private
+            return (headers[CACHE_CONTROL_HEADER].include?(NO_CACHE) ||
+                   headers[CACHE_CONTROL_HEADER].include?(NO_STORE) ||
+                   headers[CACHE_CONTROL_HEADER].include?(PRIVATE))
+          end
+
+          # last check expires
+          if headers.key?(EXPIRES_HEADER) && !headers[EXPIRES_HEADER].nil?
+            # Expires=0 means not cacced
+            return true if headers[EXPIRES_HEADER] == '0'
+          end
+        end
+
+        def get_current_trace_id
+          tracer = configuration[:tracer]
+          span = tracer.active_span
+          span.trace_id if span
+        end
+
+        def generate_updated_html(response, headers, trace_id)
+          concatted_html = concat_html_fragments(response)
+
+          unix_expiry_time = DateTime.now.strftime('%Q').to_i
+
+          meta_injection_point = find_meta_injection_point(concatted_html)
+          html_comment = html_comment_template(trace_id, unix_expiry_time)
+          meta_tag = meta_tag_template(trace_id, unix_expiry_time) if META_TAG_ENABLED
+
+          modify_html(concatted_html, html_comment, meta_injection_point, meta_tag)
+        # catch everything and swallow it here for defensiveness
+        rescue Exception => e # rubocop:disable Lint/RescueException
+          Datadog.logger.warn("Error updating html for rum injection #{e.class}: #{e.message} #{e.backtrace.join("\n")}")
+          return nil
+        end
+
+        def concat_html_fragments(response)
+          # aggregate the html into a complete document
+          # should have a max amount we parse here after which we give up
+          html_doc = nil
+          response.each do |frag|
```

Comment:
So, I thought of something that might be a little easier, which is that, since we will always be injecting the HTML Comment at the very start of the response, we can call `.each` , .insert, and then just break after doing so....since we aren't planning to support streaming afaik, this might be a little simpler than wrapping in an object and delegating methods (i'm a little hesitant to take this approach in general), and we avoid loading everything into memory like in the original logic.

What do you think?

Reply:
> we can call .each , .insert, and then just break after doing so

I'm not sure I follow it here. We should avoid, unless necessary, to call `response.each` in our middleware, as this will eagerly start loading the response. I'm not sure if there's a way to avoid calling `#each` without some form of wrapping. If your suggestion is able to fulfill this requirement, feel free to go ahead and implement it.

For wrapping, I was thinking of something like this, which [follows Rack's spec for the `body`](https://www.rubydoc.info/github/rack/rack/file/SPEC#label-The+Body):
```ruby
class RumBody
  def initialize(first_line, original_body)
    @original_body = original_body
    @new_body = Enumerator.new { |y|
      y << first_line
      original_body.each { |e| y << e }
    }
  end

  def each(&block)
    @new_body.each(&block)
  end

  def close
    @original_body.close
  end
end
```

Reply:
I've re-read the updated code, I'll leave feedback on the new implementation instead. This last comment ^ is slightly outdated, regarding my take on your approach (which I can now clearly see in code).

My suggestion above is still valid, but feedback on the implementation will be left on separate comments on the relevant lines.

Reply:
@marcotc i think its probably better to use a wrapping object...maybe using a Delegator though so that we don't have to worry about making sure all the other methods behave correctly?

Reply:
i updated this to use the wrapped Body approach. Originally i had tried using `SimpleDelegator` but was hitting some issues on old versions of ruby, so went with the approach described above, tried to match the spec behavior.